File: apps/linux-desktop/src/dashboard/homeSpaceBudget.ts
/**
 * Fit / Feed / Breathe — the living-layout solver.
 *
 * Same reasoning as macOS `HomeSpaceBudget`, native TypeScript shape. A shell
 * declares appetite (`HomeSlot`); the canvas answers with a plan
 * (`HomeSpacePlan`). Pure arithmetic: no DOM, no React, no window.
 *
 * Resolution order is the design:
 *   1. Fit — can every slot have its floor? If not, ambient furniture yields,
 *      then the surface scrolls. Content is never dropped.
 *   2. Feed — spend slack on real rows, round-robin in rank order.
 *   3. Breathe — only leftover slack becomes space.
 */

export type Size = { width: number; height: number };

export type HomeRowAppetite = {
  /** Rows the data actually has. A slot can never be fed a row that does not exist. */
  available: number;
  /** Rows shown before any slack is spent. Part of the slot's floor. */
  baseline: number;
  /** Height one row costs, including its rule. */
  unit: number;
  /** Most rows worth showing here even on an enormous canvas. */
  ceiling: number;
};

export type HomeSlot = {
  id: string;
  /** Feeding order. Lower is fed first. */
  rank: number;
  /** Height at baseline rows. Never goes below this. */
  floor: number;
  /** Height this slot would choose if nothing were competing. */
  ideal: number;
  /** Share of post-Feed slack. `0` is rigid. */
  stretch: number;
  /** More rows this slot could honestly show. Omit for a headline / field / gauge. */
  rows?: HomeRowAppetite;
  /** Withheld on a canvas too short. True only for ambient furniture. */
  isAmbient: boolean;
  /** Keeps full canvas width above any column region. Rigid at `ideal`. */
  spans: boolean;
};

export type HomePlacement = {
  id: string;
  /** Resolved height. `null` when the surface overflows and the slot should hug. */
  height: number | null;
  /** How many rows to render. `0` for slots with no appetite. */
  rowCount: number;
  /** Column the slot was dealt into, 0-based. */
  column: number;
  /** Only ambient furniture is ever withheld. */
  isVisible: boolean;
};

export type HomeSpacePlan = {
  placements: HomePlacement[];
  columns: number;
  overflows: boolean;
  spanningIDs: string[];
};

export const EMPTY_HOME_SPACE_PLAN: HomeSpacePlan = {
  placements: [],
  columns: 1,
  overflows: false,
  spanningIDs: []
};

export function createRowAppetite(input: {
  available: number;
  baseline: number;
  unit: number;
  ceiling: number;
}): HomeRowAppetite {
  const available = Math.max(0, input.available);
  return {
    available,
    baseline: Math.max(0, Math.min(input.baseline, available)),
    unit: Math.max(1, input.unit),
    ceiling: Math.max(0, input.ceiling)
  };
}

export function appetiteCap(rows: HomeRowAppetite): number {
  return Math.min(rows.available, rows.ceiling);
}

export function createHomeSlot(input: {
  id: string;
  rank: number;
  floor: number;
  ideal?: number;
  stretch?: number;
  rows?: HomeRowAppetite;
  isAmbient?: boolean;
  spans?: boolean;
}): HomeSlot {
  const floor = Math.max(0, input.floor);
  const ideal = Math.max(floor, input.ideal ?? floor);
  return {
    id: input.id,
    rank: input.rank,
    floor,
    ideal,
    stretch: Math.max(0, input.stretch ?? 0),
    rows: input.rows,
    isAmbient: input.isAmbient ?? false,
    spans: input.spans ?? false
  };
}

export function planPlacement(plan: HomeSpacePlan, id: string): HomePlacement | undefined {
  return plan.placements.find((placement) => placement.id === id);
}

export function planRowCount(plan: HomeSpacePlan, id: string, fallback = 0): number {
  return planPlacement(plan, id)?.rowCount ?? fallback;
}

export function planHeight(plan: HomeSpacePlan, id: string): number | null | undefined {
  const placement = planPlacement(plan, id);
  if (!placement) return undefined;
  return placement.height;
}

export function planIsVisible(plan: HomeSpacePlan, id: string): boolean {
  return planPlacement(plan, id)?.isVisible ?? true;
}

export function planColumnGroups(plan: HomeSpacePlan): string[][] {
  if (plan.columns <= 0) return [];
  const spanning = new Set(plan.spanningIDs);
  const groups: string[][] = [];
  for (let column = 0; column < plan.columns; column += 1) {
    groups.push(
      plan.placements
        .filter((placement) => placement.column === column && placement.isVisible && !spanning.has(placement.id))
        .map((placement) => placement.id)
    );
  }
  return groups;
}

export function planVisibleSpanningIDs(plan: HomeSpacePlan): string[] {
  return plan.spanningIDs.filter((id) => planIsVisible(plan, id));
}

type ColumnResult = {
  placements: Array<{ id: string; height: number | null; rowCount: number; isVisible: boolean }>;
  overflows: boolean;
};

function resolveColumn(slots: HomeSlot[], height: number, gutter: number): ColumnResult {
  const kept = slots.slice();
  let chrome = gutter * Math.max(0, kept.length - 1);
  let floors = kept.reduce((sum, slot) => sum + slot.floor, 0);

  while (floors + chrome > height && kept.some((slot) => slot.isAmbient)) {
    let victim = -1;
    for (let i = kept.length - 1; i >= 0; i -= 1) {
      if (kept[i]!.isAmbient) {
        victim = i;
        break;
      }
    }
    if (victim < 0) break;
    floors -= kept[victim]!.floor;
    kept.splice(victim, 1);
    chrome = gutter * Math.max(0, kept.length - 1);
  }

  const keptIDs = new Set(kept.map((slot) => slot.id));
  const withheld = slots.filter((slot) => !keptIDs.has(slot.id));

  if (floors + chrome > height) {
    return {
      placements: [
        ...kept.map((slot) => ({
          id: slot.id,
          height: null,
          rowCount: slot.rows?.baseline ?? 0,
          isVisible: true
        })),
        ...withheld.map((slot) => ({
          id: slot.id,
          height: 0,
          rowCount: 0,
          isVisible: false
        }))
      ],
      overflows: true
    };
  }

  let slack = height - floors - chrome;
  const rowCounts = new Map(kept.map((slot) => [slot.id, slot.rows?.baseline ?? 0]));
  const feeding = kept.filter((slot) => slot.rows != null).sort((a, b) => a.rank - b.rank);

  let fed = true;
  while (slack > 0 && fed) {
    fed = false;
    for (const slot of feeding) {
      const appetite = slot.rows;
      if (!appetite) continue;
      const current = rowCounts.get(slot.id) ?? 0;
      if (current >= appetiteCap(appetite)) continue;
      if (slack < appetite.unit) continue;
      rowCounts.set(slot.id, current + 1);
      slack -= appetite.unit;
      fed = true;
    }
  }

  const totalStretch = kept.reduce((sum, slot) => sum + slot.stretch, 0);
  const totalIdeal = kept.reduce((sum, slot) => sum + slot.ideal, 0);

  const placements: ColumnResult['placements'] = [];
  for (const slot of kept) {
    const rowCount = rowCounts.get(slot.id) ?? 0;
    const bought = rowCount - (slot.rows?.baseline ?? 0);
    const earned = bought * (slot.rows?.unit ?? 0);
    let share: number;
    if (totalStretch > 0) {
      share = slack * (slot.stretch / totalStretch);
    } else if (totalIdeal > 0) {
      share = slack * (slot.ideal / totalIdeal);
    } else {
      share = slack / kept.length;
    }
    placements.push({
      id: slot.id,
      height: slot.floor + earned + share,
      rowCount,
      isVisible: true
    });
  }

  for (const slot of withheld) {
    placements.push({ id: slot.id, height: 0, rowCount: 0, isVisible: false });
  }

  return { placements, overflows: false };
}

export const HomeSpaceBudget = {
  /** Width at which a composition earns a second column. */
  twoColumnWidth: 1_080,
  /** Width at which a third column earns its place. */
  threeColumnWidth: 1_580,
  /** Hysteresis either side of a threshold, so a drag cannot flicker. */
  columnDeadBand: 60,

  columns(width: number, current: number, slots: number): number {
    if (width <= 0) return Math.max(1, Math.min(current, Math.max(1, slots)));

    const ceiling = Math.max(1, Math.min(3, slots));
    let target: number;
    if (width >= this.threeColumnWidth + this.columnDeadBand) {
      target = 3;
    } else if (width <= this.threeColumnWidth - this.columnDeadBand) {
      if (width >= this.twoColumnWidth + this.columnDeadBand) {
        target = 2;
      } else if (width <= this.twoColumnWidth - this.columnDeadBand) {
        target = 1;
      } else {
        target = Math.min(Math.max(current, 1), 2);
      }
    } else {
      target = Math.min(Math.max(current, 2), 3);
    }
    return Math.min(target, ceiling);
  },

  deal(slots: HomeSlot[], columns: number): Record<string, number> {
    if (columns <= 1) {
      return Object.fromEntries(slots.map((slot) => [slot.id, 0]));
    }

    const loads = Array.from({ length: columns }, () => 0);
    const assignment: Record<string, number> = {};
    const ranked = slots.slice().sort((a, b) => a.rank - b.rank);

    for (const slot of ranked) {
      let target = 0;
      for (let column = 1; column < columns; column += 1) {
        if (loads[column]! < loads[target]! - 0.5) target = column;
      }
      assignment[slot.id] = target;
      loads[target]! += slot.ideal;
    }
    return assignment;
  },

  resolve(input: {
    canvas: Size;
    slots: HomeSlot[];
    gutter: number;
    columns?: number;
  }): HomeSpacePlan {
    const { canvas, slots, gutter } = input;
    if (slots.length === 0) return EMPTY_HOME_SPACE_PLAN;

    const requested = Math.max(1, input.columns ?? 1);
    const spanning = requested > 1 ? slots.filter((slot) => slot.spans) : [];
    const columnar = requested > 1 ? slots.filter((slot) => !slot.spans) : slots;
    const spanningIDs = spanning.map((slot) => slot.id);
    const columnCount = Math.max(1, Math.min(requested, Math.max(1, columnar.length)));
    const assignment = this.deal(columnar, columnCount);

    if (canvas.height <= 0) {
      return {
        placements: slots.map((slot) => ({
          id: slot.id,
          height: null,
          rowCount: slot.rows?.baseline ?? 0,
          column: assignment[slot.id] ?? 0,
          isVisible: true
        })),
        columns: columnCount,
        overflows: true,
        spanningIDs
      };
    }

    const placements: HomePlacement[] = [];
    let anyColumnOverflows = false;

    const bandHeight = spanning.reduce((sum, slot) => sum + slot.ideal, 0) + gutter * spanning.length;
    const columnHeight = canvas.height - bandHeight;

    for (const slot of spanning) {
      placements.push({
        id: slot.id,
        height: slot.ideal,
        rowCount: slot.rows?.baseline ?? 0,
        column: 0,
        isVisible: true
      });
    }

    for (let column = 0; column < columnCount; column += 1) {
      const members = columnar.filter((slot) => (assignment[slot.id] ?? 0) === column);
      if (members.length === 0) continue;
      const resolved = resolveColumn(members, columnHeight, gutter);
      anyColumnOverflows = anyColumnOverflows || resolved.overflows;
      for (const placement of resolved.placements) {
        placements.push({
          id: placement.id,
          height: placement.height,
          rowCount: placement.rowCount,
          column,
          isVisible: placement.isVisible
        });
      }
    }

    let finalPlacements = placements;
    if (anyColumnOverflows) {
      finalPlacements = placements.map((placement) => ({ ...placement, height: null }));
    }

    const order = new Map(slots.map((slot, index) => [slot.id, index]));
    finalPlacements.sort((a, b) => (order.get(a.id) ?? 0) - (order.get(b.id) ?? 0));

    return {
      placements: finalPlacements,
      columns: columnCount,
      overflows: anyColumnOverflows,
      spanningIDs
    };
  }
};

File: windows/app/OpenBurnBar.App.Dashboard/Layout/HomeSpaceBudget.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace OpenBurnBar.App.Dashboard.Layout;

/// <summary>
/// One region a shell asks the canvas for.
/// A slot describes appetite, never geometry: what it cannot go below, what it
/// would like, how many more rows it could honestly fill, and how willing it is
/// to absorb leftover height. The canvas decides the numbers.
/// </summary>
public sealed class HomeSlot : IEquatable<HomeSlot>
{
    /// <summary>
    /// How much more real content a slot could show if the canvas can afford it.
    /// A slot with an appetite grows by revealing rows that already exist in the data
    /// and were being truncated; a slot without one grows by stretching.
    /// </summary>
    public sealed class RowAppetite : IEquatable<RowAppetite>
    {
        public int Available { get; }
        public int Baseline { get; }
        public double Unit { get; }
        public int Ceiling { get; }

        public RowAppetite(int available, int baseline, double unit, int ceiling)
        {
            Available = Math.Max(0, available);
            Baseline = Math.Max(0, Math.Min(baseline, Math.Max(0, available)));
            Unit = Math.Max(1.0, unit);
            Ceiling = Math.Max(0, ceiling);
        }

        /// <summary>The most rows this appetite can ever resolve to.</summary>
        public int Cap => Math.Min(Available, Ceiling);

        public bool Equals(RowAppetite? other)
        {
            if (other is null) return false;
            if (ReferenceEquals(this, other)) return true;
            return Available == other.Available &&
                   Baseline == other.Baseline &&
                   Math.Abs(Unit - other.Unit) < 0.0001 &&
                   Ceiling == other.Ceiling;
        }

        public override bool Equals(object? obj) => Equals(obj as RowAppetite);

        public override int GetHashCode() => HashCode.Combine(Available, Baseline, Unit, Ceiling);
    }

    public string Id { get; }
    public int Rank { get; }
    public double Floor { get; }
    public double Ideal { get; }
    public double Stretch { get; }
    public RowAppetite? Rows { get; }
    public bool IsAmbient { get; }
    public bool Spans { get; }

    public HomeSlot(
        string id,
        int rank = 0,
        double floor = 100,
        double? ideal = null,
        double stretch = 0,
        RowAppetite? rows = null,
        bool isAmbient = false,
        bool spans = false)
    {
        Id = id ?? throw new ArgumentNullException(nameof(id));
        Rank = rank;
        Floor = Math.Max(0, floor);
        Ideal = Math.Max(Math.Max(0, floor), ideal ?? floor);
        Stretch = Math.Max(0, stretch);
        Rows = rows;
        IsAmbient = isAmbient;
        Spans = spans;
    }

    public bool Equals(HomeSlot? other)
    {
        if (other is null) return false;
        if (ReferenceEquals(this, other)) return true;
        return Id == other.Id &&
               Rank == other.Rank &&
               Math.Abs(Floor - other.Floor) < 0.0001 &&
               Math.Abs(Ideal - other.Ideal) < 0.0001 &&
               Math.Abs(Stretch - other.Stretch) < 0.0001 &&
               Equals(Rows, other.Rows) &&
               IsAmbient == other.IsAmbient &&
               Spans == other.Spans;
    }

    public override bool Equals(object? obj) => Equals(obj as HomeSlot);

    public override int GetHashCode() => HashCode.Combine(Id, Rank, Floor, Ideal, Stretch, Rows, IsAmbient, Spans);
}

/// <summary>
/// What the canvas granted, ready to be rendered without further arithmetic.
/// </summary>
public sealed class HomeSpacePlan : IEquatable<HomeSpacePlan>
{
    public sealed class Placement : IEquatable<Placement>
    {
        public string Id { get; }
        public double? Height { get; }
        public int RowCount { get; }
        public int Column { get; }
        public bool IsVisible { get; }

        public Placement(string id, double? height, int rowCount, int column, bool isVisible)
        {
            Id = id ?? throw new ArgumentNullException(nameof(id));
            Height = height;
            RowCount = rowCount;
            Column = column;
            IsVisible = isVisible;
        }

        public bool Equals(Placement? other)
        {
            if (other is null) return false;
            if (ReferenceEquals(this, other)) return true;
            bool heightEq = (!Height.HasValue && !other.Height.HasValue) ||
                            (Height.HasValue && other.Height.HasValue && Math.Abs(Height.Value - other.Height.Value) < 0.001);
            return Id == other.Id &&
                   heightEq &&
                   RowCount == other.RowCount &&
                   Column == other.Column &&
                   IsVisible == other.IsVisible;
        }

        public override bool Equals(object? obj) => Equals(obj as Placement);

        public override int GetHashCode() => HashCode.Combine(Id, Height, RowCount, Column, IsVisible);
    }

    public IReadOnlyList<Placement> Placements { get; }
    public int Columns { get; }
    public bool Overflows { get; }
    public IReadOnlyList<string> SpanningIDs { get; }

    public HomeSpacePlan(
        IReadOnlyList<Placement> placements,
        int columns,
        bool overflows,
        IReadOnlyList<string> spanningIDs)
    {
        Placements = placements ?? Array.Empty<Placement>();
        Columns = columns;
        Overflows = overflows;
        SpanningIDs = spanningIDs ?? Array.Empty<string>();
    }

    public Placement? GetPlacement(string id) =>
        Placements.FirstOrDefault(p => string.Equals(p.Id, id, StringComparison.Ordinal));

    public int RowCount(string id, int fallback = 0) =>
        GetPlacement(id)?.RowCount ?? fallback;

    public double? Height(string id) =>
        GetPlacement(id)?.Height;

    public bool IsVisible(string id) =>
        GetPlacement(id)?.IsVisible ?? true;

    public IReadOnlyList<IReadOnlyList<string>> ColumnGroups
    {
        get
        {
            if (Columns <= 0) return Array.Empty<IReadOnlyList<string>>();
            var spanningSet = new HashSet<string>(SpanningIDs, StringComparer.Ordinal);
            var groups = new List<IReadOnlyList<string>>(Columns);
            for (int col = 0; col < Columns; col++)
            {
                int c = col;
                var members = Placements
                    .Where(p => p.Column == c && p.IsVisible && !spanningSet.Contains(p.Id))
                    .Select(p => p.Id)
                    .ToList();
                groups.Add(members);
            }
            return groups;
        }
    }

    public IReadOnlyList<string> VisibleSpanningIDs =>
        SpanningIDs.Where(IsVisible).ToList();

    public static readonly HomeSpacePlan Empty = new(
        Array.Empty<Placement>(),
        columns: 1,
        overflows: false,
        Array.Empty<string>());

    public bool Equals(HomeSpacePlan? other)
    {
        if (other is null) return false;
        if (ReferenceEquals(this, other)) return true;
        if (Columns != other.Columns || Overflows != other.Overflows) return false;
        if (Placements.Count != other.Placements.Count || SpanningIDs.Count != other.SpanningIDs.Count) return false;
        for (int i = 0; i < Placements.Count; i++)
        {
            if (!Placements[i].Equals(other.Placements[i])) return false;
        }
        for (int i = 0; i < SpanningIDs.Count; i++)
        {
            if (!string.Equals(SpanningIDs[i], other.SpanningIDs[i], StringComparison.Ordinal)) return false;
        }
        return true;
    }

    public override bool Equals(object? obj) => Equals(obj as HomeSpacePlan);

    public override int GetHashCode() => HashCode.Combine(Placements.Count, Columns, Overflows, SpanningIDs.Count);
}

/// <summary>
/// Standard reading width measures (prose properties, not page caps).
/// </summary>
public static class HomeReadingMeasure
{
    public const double Headline = 620;
    public const double Body = 680;
    public const double Editorial = 760;
}

/// <summary>
/// Resolves slot appetite against a canvas via Fit, Feed, Breathe passes.
/// Pure arithmetic on values; matches macOS HomeSpaceBudget.swift.
/// </summary>
public static class HomeSpaceBudget
{
    public const double TwoColumnWidth = 1080;
    public const double ThreeColumnWidth = 1580;
    public const double ColumnDeadBand = 60;

    /// <summary>
    /// How many columns this width supports, holding the current count inside the
    /// dead band so a drag across a threshold cannot flicker.
    /// </summary>
    public static int Columns(double width, int current, int slots)
    {
        if (width <= 0)
        {
            return Math.Max(1, Math.Min(current, Math.Max(1, slots)));
        }

        int ceiling = Math.Max(1, Math.Min(3, slots));
        int target;
        if (width >= ThreeColumnWidth + ColumnDeadBand)
        {
            target = 3;
        }
        else if (width <= ThreeColumnWidth - ColumnDeadBand)
        {
            if (width >= TwoColumnWidth + ColumnDeadBand)
            {
                target = 2;
            }
            else if (width <= TwoColumnWidth - ColumnDeadBand)
            {
                target = 1;
            }
            else
            {
                target = Math.Min(Math.Max(current, 1), 2);
            }
        }
        else
        {
            target = Math.Min(Math.Max(current, 2), 3);
        }

        return Math.Min(target, ceiling);
    }

    /// <summary>
    /// Deals slots into columns, balancing by ideal height.
    /// Rank order into the currently shortest column, ties go left.
    /// </summary>
    public static Dictionary<string, int> Deal(IReadOnlyList<HomeSlot> slots, int columns)
    {
        if (columns <= 1)
        {
            return slots.ToDictionary(s => s.Id, _ => 0, StringComparer.Ordinal);
        }

        var loads = new double[columns];
        var assignment = new Dictionary<string, int>(StringComparer.Ordinal);

        foreach (HomeSlot slot in slots.OrderBy(s => s.Rank))
        {
            int target = 0;
            for (int col = 1; col < columns; col++)
            {
                if (loads[col] < loads[target] - 0.5)
                {
                    target = col;
                }
            }
            assignment[slot.Id] = target;
            loads[target] += slot.Ideal;
        }

        return assignment;
    }

    /// <summary>
    /// Resolves slots against a canvas using Fit, Feed, Breathe.
    /// </summary>
    public static HomeSpacePlan Resolve(
        double canvasWidth,
        double canvasHeight,
        IReadOnlyList<HomeSlot> slots,
        double gutter,
        int requestedColumns = 1)
    {
        if (slots == null || slots.Count == 0)
        {
            return HomeSpacePlan.Empty;
        }

        int requested = Math.Max(1, requestedColumns);
        List<HomeSlot> spanning = requested > 1
            ? slots.Where(s => s.Spans).ToList()
            : new List<HomeSlot>();
        List<HomeSlot> columnar = requested > 1
            ? slots.Where(s => !s.Spans).ToList()
            : slots.ToList();
        List<string> spanningIDs = spanning.Select(s => s.Id).ToList();

        int columnCount = Math.Max(1, Math.Min(requested, Math.Max(1, columnar.Count)));
        Dictionary<string, int> assignment = Deal(columnar, columnCount);

        if (canvasHeight <= 0)
        {
            var zeroPlacements = slots.Select(slot => new HomeSpacePlan.Placement(
                id: slot.Id,
                height: null,
                rowCount: slot.Rows?.Baseline ?? 0,
                column: assignment.TryGetValue(slot.Id, out int col) ? col : 0,
                isVisible: true
            )).ToList();

            return new HomeSpacePlan(
                placements: zeroPlacements,
                columns: columnCount,
                overflows: true,
                spanningIDs: spanningIDs);
        }

        var placements = new List<HomeSpacePlan.Placement>();
        bool anyColumnOverflows = false;

        double bandHeight = spanning.Sum(s => s.Ideal) + gutter * spanning.Count;
        double columnHeight = canvasHeight - bandHeight;

        foreach (HomeSlot slot in spanning)
        {
            placements.Add(new HomeSpacePlan.Placement(
                id: slot.Id,
                height: slot.Ideal,
                rowCount: slot.Rows?.Baseline ?? 0,
                column: 0,
                isVisible: true));
        }

        for (int col = 0; col < columnCount; col++)
        {
            int c = col;
            var members = columnar.Where(s => assignment.TryGetValue(s.Id, out int colIdx) && colIdx == c).ToList();
            if (members.Count == 0) continue;

            ColumnResult resolved = ResolveColumn(members, columnHeight, gutter);
            anyColumnOverflows = anyColumnOverflows || resolved.Overflows;
            foreach (var p in resolved.Placements)
            {
                placements.Add(new HomeSpacePlan.Placement(
                    id: p.Id,
                    height: p.Height,
                    rowCount: p.RowCount,
                    column: c,
                    isVisible: p.IsVisible));
            }
        }

        if (anyColumnOverflows)
        {
            placements = placements.Select(p => new HomeSpacePlan.Placement(
                id: p.Id,
                height: p.IsVisible ? null : 0,
                rowCount: p.RowCount,
                column: p.Column,
                isVisible: p.IsVisible)).ToList();
        }

        var order = slots.Select((s, idx) => (s.Id, idx))
                         .ToDictionary(x => x.Id, x => x.idx, StringComparer.Ordinal);
        placements.Sort((a, b) =>
        {
            int orderA = order.TryGetValue(a.Id, out int idxA) ? idxA : 0;
            int orderB = order.TryGetValue(b.Id, out int idxB) ? idxB : 0;
            return orderA.CompareTo(orderB);
        });

        return new HomeSpacePlan(
            placements: placements,
            columns: columnCount,
            overflows: anyColumnOverflows,
            spanningIDs: spanningIDs);
    }

    private sealed class ColumnResult
    {
        public IReadOnlyList<(string Id, double? Height, int RowCount, bool IsVisible)> Placements { get; }
        public bool Overflows { get; }

        public ColumnResult(
            IReadOnlyList<(string Id, double? Height, int RowCount, bool IsVisible)> placements,
            bool overflows)
        {
            Placements = placements;
            Overflows = overflows;
        }
    }

    private static ColumnResult ResolveColumn(
        IReadOnlyList<HomeSlot> slots,
        double height,
        double gutter)
    {
        var kept = slots.ToList();
        double chrome = gutter * Math.Max(0, kept.Count - 1);
        double floors = kept.Sum(s => s.Floor);

        // Fit pass — ambient furniture yields first
        while (floors + chrome > height && kept.Any(s => s.IsAmbient))
        {
            int victim = kept.FindLastIndex(s => s.IsAmbient);
            if (victim < 0) break;
            floors -= kept[victim].Floor;
            kept.RemoveAt(victim);
            chrome = gutter * Math.Max(0, kept.Count - 1);
        }

        var keptIDs = new HashSet<string>(kept.Select(s => s.Id), StringComparer.Ordinal);
        var withheld = slots.Where(s => !keptIDs.Contains(s.Id)).ToList();

        if (floors + chrome > height)
        {
            var overflowPlacements = kept.Select(s => (s.Id, (double?)null, s.Rows?.Baseline ?? 0, true))
                .Concat(withheld.Select(s => (s.Id, (double?)0.0, 0, false)))
                .ToList();
            return new ColumnResult(overflowPlacements, overflows: true);
        }

        double slack = height - floors - chrome;
        var rowCounts = kept.ToDictionary(s => s.Id, s => s.Rows?.Baseline ?? 0, StringComparer.Ordinal);

        // Feed pass — round-robin across slots in rank order
        var feeding = kept.Where(s => s.Rows != null).OrderBy(s => s.Rank).ToList();
        bool fed = true;
        while (slack > 0 && fed)
        {
            fed = false;
            foreach (HomeSlot slot in feeding)
            {
                HomeSlot.RowAppetite appetite = slot.Rows!;
                if (rowCounts.TryGetValue(slot.Id, out int current) && current < appetite.Cap)
                {
                    if (slack >= appetite.Unit)
                    {
                        rowCounts[slot.Id] = current + 1;
                        slack -= appetite.Unit;
                        fed = true;
                    }
                }
            }
        }

        // Breathe pass — leftover slack becomes space
        double totalStretch = kept.Sum(s => s.Stretch);
        double totalIdeal = kept.Sum(s => s.Ideal);

        var resultPlacements = new List<(string Id, double? Height, int RowCount, bool IsVisible)>();
        foreach (HomeSlot slot in kept)
        {
            int currentRows = rowCounts[slot.Id];
            int baseline = slot.Rows?.Baseline ?? 0;
            int bought = currentRows - baseline;
            double earned = bought * (slot.Rows?.Unit ?? 0);

            double share;
            if (totalStretch > 0)
            {
                share = slack * (slot.Stretch / totalStretch);
            }
            else if (totalIdeal > 0)
            {
                share = slack * (slot.Ideal / totalIdeal);
            }
            else
            {
                share = slack / kept.Count;
            }

            resultPlacements.Add((slot.Id, slot.Floor + earned + share, currentRows, true));
        }

        foreach (HomeSlot slot in withheld)
        {
            resultPlacements.Add((slot.Id, 0.0, 0, false));
        }

        return new ColumnResult(resultPlacements, overflows: false);
    }
}
